src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs

```diff
@@ -317,8 +317,8 @@ protected HqlTreeNode VisitBinaryExpression(BinaryExpression expression)
 			var rightType = GetExpressionType(expression.Right);
 			if (leftType != null && leftType == rightType)
 			{
-				_notCastableExpressions.Add(expression.Left, leftType);
-				_notCastableExpressions.Add(expression.Right, rightType);
+				_notCastableExpressions.TryAdd(expression.Left, leftType);
```

Comment:
TryAdd is not available in .net standard 2.0 and .net 4.6.1

Reply:
Added an extension method for them.

Reply:
@maca88 can we replace Add with Set? 

_notCastableExpressions[expression.Right] = rightType;

Reply:
Somehow I overlooked the comment. Yes we can, replaced with Set. 
Also now I realized that I created the branch to this repository instead on my fork. We should remove the branch after merging.

Reply:
>Somehow I overlooked the comment.

This is what I thought.